src/future/flatten_sink.rs

```diff
@@ -0,0 +1,91 @@
+use std::mem;
+
+use {Async, AsyncSink, StartSend, Poll, Future, Sink};
+
+
+enum State<F: Future>
+    where F::Item: Sink
+{
+    Future(F),
+    Sink(F::Item),
+    Error(<F::Item as Sink>::SinkError),
+    Void,
+}
+
+/// Future for the `into_sink` combinator, flattening a future-of-a-sink to
+/// get just the result of the final sink instead
+///
+/// This is created by the `Future::into_sink` method.
+pub struct FlattenSink<F: Future>
+    where F::Item: Sink
+{
+    state: State<F>,
+}
+
+pub fn new<F>(future: F) -> FlattenSink<F>
+    where F: Future,
+          F::Item: Sink
+{
+    FlattenSink {
+        state: State::Future(future),
+    }
+}
+
+impl<F: Future> Sink for FlattenSink<F>
+    where F::Item: Sink,
+          <F::Item as Sink>::SinkError: From<F::Error>,
+{
+    type SinkItem = <F::Item as Sink>::SinkItem;
+    type SinkError = <F::Item as Sink>::SinkError;
+    fn start_send(&mut self, item: Self::SinkItem)
+        -> StartSend<Self::SinkItem, Self::SinkError>
+    {
+        let (res, state) = match mem::replace(&mut self.state, State::Void) {
+            State::Future(mut fut) => {
+                match fut.poll() {
+                    Ok(Async::Ready(mut sink)) => {
+                        (sink.start_send(item)?, State::Sink(sink))
```

Comment:
Well, I'll fix the code if we agree on that, but your previous message was confusing :). So I need to make things more clear:

> I do think though that if the future returns the error then this implementation should just be fused to forever return NotReady

Why should it be fused?

1. Most futures panic when poll'ed after error
2. Sink is documented that can panic when polled after error

I think it's more clear to panic, because there is no real use case of the fused sink. And calling poll_complete/start_send again means there is a bug in code that silenced an error.

Even if we do return `NotReady` on the error of the future user still has to differentiate the error generated by Future and error generated by Sink because there is no sense to continue using sink when Future returned error (it just *hangs* from the viewpoint of the user).

And I think this part:
> I figure that we'd delegate to the underlying sink to "act improperly", if at all

.. only works if we return `NotReady` after errors of the Future, right?